numpy/core/src/multiarray/item_selection.c

```diff
@@ -1749,6 +1750,275 @@ PyArray_LexSort(PyObject *sort_keys, int axis)
+                                       PyArray_NDIM(mps[0])))) {
+                PyErr_SetString(PyExc_ValueError,
+                                "all keys need to be the same shape");
+                goto fail;
+            }
+        }
+        if (!PyArray_DESCR(mps[i])->f->argsort[NPY_STABLESORT]
+                && !PyArray_DESCR(mps[i])->f->compare) {
+            PyErr_Format(PyExc_TypeError,
+                         "item %zd type does not have compare function", i);
+            goto fail;
+        }
+        if (!object
+            && PyDataType_FLAGCHK(PyArray_DESCR(mps[i]), NPY_NEEDS_PYAPI)) {
+            object = 1;
+        }
+        if ((PyArray_NDIM(mps[i]) != PyArray_NDIM(self))||
+            (!PyArray_CompareLists(PyArray_DIMS(mps[i]),
+                                       PyArray_DIMS(self),
+                                       PyArray_NDIM(self)))) {
+                PyErr_SetString(PyExc_ValueError,
+                                "all keys and input array need to be the same shape");
+                goto fail;
+        }
+    }
+
+    /* Now we can check the axis */
+    nd = PyArray_NDIM(mps[0]);
+    /*
+    * Special case letting axis={-1,0} slip through for scalars,
+    * for backwards compatibility reasons.
+    */
+    if (nd == 0 && (axis == 0 || axis == -1)) {
+        /* TODO: can we deprecate this? */
+    }
+    else if (check_and_adjust_axis(&axis, nd) < 0) {
+        goto fail;
+    }
+    if ((nd == 0) || (PyArray_SIZE(mps[0]) <= 1)) {
+        /* empty/single element case */
+        // Nothing to do here just return
+        printf("Should be doing nothing\n");
+        fflush(stdout);
+        goto finish;
+    }
+
+    for (i = 0; i < n; i++) {
+        its[i] = (PyArrayIterObject *)PyArray_IterAllButAxis(
+                (PyObject *)mps[i], &axis);
+        if (its[i] == NULL) {
+            goto fail;
+        }
+    }
+    /* Now do the sorting */
+    sit = (PyArrayIterObject *)
+            PyArray_IterAllButAxis((PyObject *)self, &axis);
+    if (sit == NULL) {
+        goto fail;
+    }
+    if (!object) {
+        NPY_BEGIN_THREADS;
+    }
+    size = sit->size;
+    // Number of elements in a slice is denoted by N
+    N = PyArray_DIMS(mps[0])[axis];
+    self_stride = PyArray_STRIDE(self, axis);
+    self_elsize = PyArray_DESCR(self)->elsize;
+    maxelsize = PyArray_DESCR(mps[0])->elsize;
+    needcopy = (self_stride != self_elsize);
+    for (j = 0; j < n; j++) {
+        needcopy = needcopy
+            || PyArray_ISBYTESWAPPED(mps[j])
+            || !(PyArray_FLAGS(mps[j]) & NPY_ARRAY_ALIGNED)
+            || (PyArray_STRIDES(mps[j])[axis] != (npy_intp)PyArray_DESCR(mps[j])->elsize);
+        if (PyArray_DESCR(mps[j])->elsize > maxelsize) {
+            maxelsize = PyArray_DESCR(mps[j])->elsize;
+        }
+    }
+    assert(N > 0);  /* Guaranteed and assumed by indbuffer */
+    char *indbuffer = PyDataMem_NEW(N * sizeof(npy_intp));
+    if (indbuffer == NULL) {
```

Comment:
Everywhere you use this you just end up casting it. Why not:
```suggestion
    npy_intp *indbuffer = PyDataMem_NEW(N * sizeof(*indbuffer));
    if (indbuffer == NULL) {
```